### PR TITLE
Carpark

### PR DESCRIPTION
**Pull request by WBGroupR**
Hi Guys,

Finally, here is my pull request for the carpark module.

Wayne

### CODE REVIEW

**Review by navajo00 — APPROVED** [review]
Code Style: The code style complies with the Google Java Style Guide. 
Recommendation: Some code might need some comments to improve its readability.

**Comment by duane-mcmahon** [comment]
ok, it has been reviewed, i will merge.

File: src/bcccp/carpark/Carpark.java [hunk 1]
```diff
@@ -1,150 +1,170 @@
...
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Date;
 
+import bcccp.tickets.adhoc.AdhocTicketDAO;
 import bcccp.tickets.adhoc.IAdhocTicket;
 import bcccp.tickets.adhoc.IAdhocTicketDAO;
 import bcccp.tickets.season.ISeasonTicket;
 import bcccp.tickets.season.ISeasonTicketDAO;
+import bcccp.tickets.season.SeasonTicketDAO;
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:13]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 2]
```diff
@@ -1,150 +1,170 @@
...
 
 public class Carpark implements ICarpark {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:14]: SECURITY. Fundamental: Make class final if not being used for inheritance.

File: src/bcccp/carpark/Carpark.java [hunk 3]
```diff
@@ -1,150 +1,170 @@
...
-	
-	private List<ICarparkObserver> observers;
-	private String carparkId;
-	private int capacity;
-	private int numberOfCarsParked;
-	private IAdhocTicketDAO adhocTicketDAO;
-	private ISeasonTicketDAO seasonTicketDAO;
-
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:18]: FORMATING: Source file structure: Exactly one blank line separates each section that is present in the source code.

File: src/bcccp/carpark/Carpark.java [hunk 4]
```diff
@@ -1,150 +1,170 @@
...
-			ISeasonTicketDAO seasonTicketDAO) {
-		//TODO Implement constructor
-	}
 
+    private List<ICarparkObserver> observers;
+    private String carparkId;
+    private int capacity;
+    private int numberOfCarsParked;
+    private IAdhocTicketDAO adhocTicketDAO;
+    private ISeasonTicketDAO seasonTicketDAO;
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:23]: FORMATING: Source file structure: Exactly one blank line separates each section that is present in the source code.

File: src/bcccp/carpark/Carpark.java [hunk 5]
```diff
@@ -1,150 +1,170 @@
...
+    public void register(ICarparkObserver observer) {
 
+        observers.add(observer);
 
-	@Override
-	public String getName() {
-		// TODO Auto-generated method stub
-		return null;
-	}
+    }
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:43]: FORMATING: Source file structure: Exactly one blank line separates each section that is present in the source code.

File: src/bcccp/carpark/Carpark.java [hunk 6]
```diff
@@ -1,150 +1,170 @@
...
-		// TODO Auto-generated method stub
-		
-	}
+    @Override
+    public void deregisterSeasonTicket(ISeasonTicket seasonTicket) {
+
+        seasonTicketDAO.deregisterTicket(seasonTicket);
+
+    }
+
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:128]: FORMATING: Source file structure: Exactly one blank line separates each section that is present in the source code.

File: src/bcccp/carpark/Carpark.java [hunk 7]
```diff
@@ -1,150 +1,170 @@
...
+        Date dateTime = new Date();
+
+        ISeasonTicket sTicket = seasonTicketDAO.findTicketById(ticketId);
+
+        return ((dateTime.getTime() >= sTicket.getStartValidPeriod()) &&
+                (dateTime.getTime() <= sTicket.getEndValidPeriod()))
+                ? true : false;
+    }
+
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:143]: FORMATING: Source file structure: Exactly one blank line separates each section that is present in the source code.

File: src/bcccp/carpark/Carpark.java [hunk 8]
```diff
@@ -1,150 +1,170 @@
...
+
+    @Override
+    public boolean isSeasonTicketInUse(String ticketId) {
+
+        ISeasonTicket sTicket = seasonTicketDAO.findTicketById(ticketId);
+
+        return sTicket.inUse();
+    }
+
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:152]: FORMATING: Source file structure: Exactly one blank line separates each section that is present in the source code.

File: src/bcccp/carpark/Carpark.java [hunk 9]
```diff
@@ -1,150 +1,170 @@
...
+
+    @Override
+    public void recordSeasonTicketEntry(String ticketId) {
+
+        seasonTicketDAO.recordTicketEntry(ticketId);
+
+    }
+
+
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:161]: FORMATING: Source file structure: Exactly one blank line separates each section that is present in the source code.

File: src/bcccp/carpark/Carpark.java [hunk 10]
```diff
@@ -1,150 +1,170 @@
...
+    private IAdhocTicketDAO adhocTicketDAO;
+    private ISeasonTicketDAO seasonTicketDAO;
 
 
-	@Override
-	public void register(ICarparkObserver observer) {
-		// TODO Auto-generated method stub
-		
-	}
+    public Carpark(String name, int capacity,
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:24]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 11]
```diff
@@ -1,150 +1,170 @@
...
+        observers = new ArrayList<>();
+    }
 
-	@Override
-	public void deregister(ICarparkObserver observer) {
-		// TODO Auto-generated method stub
-		
-	}
 
+    @Override
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:37]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 12]
```diff
@@ -1,150 +1,170 @@
...
-	@Override
-	public String getName() {
-		// TODO Auto-generated method stub
-		return null;
-	}
+    }
 
 
+    @Override
+    public void deregister(ICarparkObserver observer) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:46]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 13]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
-	@Override
-	public void recordAdhocTicketExit() {
-		// TODO Auto-generated method stub
-		
-	}
 
+    @Override
+    public IAdhocTicket getAdhocTicket(String barcode) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:84]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 14]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
 
-	@Override
-	public float calculateAddHocTicketCharge(long entryDateTime) {
-		// TODO Auto-generated method stub
-		return 0;
-	}
+    @Override
+    public void recordAdhocTicketEntry() {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:76]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 15]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
 
-	@Override
-	public IAdhocTicket getAdhocTicket(String barcode) {
-		// TODO Auto-generated method stub
-		return null;
-	}
+    @Override
+    public IAdhocTicket issueAdhocTicket() {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:69]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 16]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
 
-	@Override
-	public void recordAdhocTicketEntry() {
-		// TODO Auto-generated method stub
-		
-	}
+    @Override
+    public boolean isFull() {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:62]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 17]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
 
-	@Override
-	public IAdhocTicket issueAdhocTicket() {
-		// TODO Auto-generated method stub
-		return null;
-	}
+    @Override
+    public String getName() {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:55]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 18]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
-	@Override
-	public void registerSeasonTicket(ISeasonTicket seasonTicket) {
-		// TODO Auto-generated method stub
-		
-	}
 
+    @Override
+    public float calculateAddHocTicketCharge(long entryDateTime) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:91]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 19]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
-	@Override
-	public boolean isSeasonTicketValid(String ticketId) {
-		// TODO Auto-generated method stub
-		return false;
-	}
 
+    @Override
+    public void recordAdhocTicketExit() {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:107]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 20]
```diff
@@ -1,150 +1,170 @@
...
-	@Override
-	public boolean isSeasonTicketInUse(String ticketId) {
-		// TODO Auto-generated method stub
-		return false;
-	}
+    }
 
 
+    @Override
+    public void registerSeasonTicket(ISeasonTicket seasonTicket) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:115]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 21]
```diff
@@ -1,150 +1,170 @@
...
+    }
 
 
-	@Override
-	public void recordSeasonTicketExit(String ticketId) {
-		// TODO Auto-generated method stub
-		
-	}
+    @Override
+    public void deregisterSeasonTicket(ISeasonTicket seasonTicket) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:123]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 22]
```diff
@@ -1,150 +1,170 @@
...
+
+        seasonTicketDAO.deregisterTicket(seasonTicket);
+
+    }
+
+
+    @Override
+    public boolean isSeasonTicketValid(String ticketId) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:131]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 23]
```diff
@@ -1,150 +1,170 @@
...
+    @Override
+    public void recordSeasonTicketEntry(String ticketId) {
+
+        seasonTicketDAO.recordTicketEntry(ticketId);
+
+    }
+
+
+    @Override
+    public void recordSeasonTicketExit(String ticketId) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:163]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

File: src/bcccp/carpark/Carpark.java [hunk 24]
```diff
@@ -1,150 +1,170 @@
...
+    public boolean isSeasonTicketInUse(String ticketId) {
+
+        ISeasonTicket sTicket = seasonTicketDAO.findTicketById(ticketId);
+
+        return sTicket.inUse();
+    }
+
+
+    @Override
+    public void recordSeasonTicketEntry(String ticketId) {
```
Comment (navajo00) [inline src/bcccp/carpark/Carpark.java:155]: JAVADOC: Please add javadoc as per Google Java Style guide.
At the minimum, Javadoc is present for every public class, and every public or protected member of such a class.

**Review by navajo00 — COMMENTED** [review]
FORMATING: Clean Code. 
*Block indentation as per: "Each time a new block or block-like construct is opened, the indent increases by two spaces. When the block ends, the indent returns to the previous indent level. The indent level applies to both code and comments throughout the block."
*There is no duplication of code. No changes required.
* Braces correctly used.
*One statement per line and a column limit of 100 characters.
!!Changes required to fix black lines among the code.

NAMING: Correct use of Java naming conventions.
Formatting of local variable names is written in lowerCamelCase.
No changes required.

JAVADOC: Changes required. Please see the comments.

Source code clean and readable. Use of getters/setter methods correctly used.

**Pull request merged**